direct/core/src/main/java/cz/o2/proxima/direct/transaction/TransactionalOnlineAttributeWriter.java

```diff
@@ -204,23 +276,55 @@ private Response takeResponse() {
             if (applicableTransforms != null) {
               applicableTransforms
                   .stream()
+                  .filter(t -> !(t.getTransformation() instanceof TransactionValidator))
                   .filter(t -> t.getFilter().apply(el))
-                  .forEach(
-                      td ->
-                          td.getTransformation()
-                              .asElementWiseTransform()
-                              .apply(
-                                  el,
-                                  transformed ->
-                                      newElements.add(injectSequenceIdAndStamp(transformed))));
+                  .forEach(td -> applyTransform(newElements, el, td));
             }
           }
         }
         currentElements = newElements;
       } while (!currentElements.isEmpty());
+      applyValidations(elements);
       return elements;
     }
 
+    private void applyValidations(Set<StreamElement> elements) {
+      for (StreamElement el : elements) {
+        List<TransformationDescriptor> applicableTransforms =
+            attributeTransforms.get(el.getAttributeDescriptor());
+        if (applicableTransforms != null) {
+          applicableTransforms
+              .stream()
+              .filter(t -> t.getTransformation() instanceof TransactionValidator)
```

Comment:
nit: both this instanceof checks looks weird for me. I understand why you do it but it lt can bring a lot of hidden issues. Maybe we should add something like `default bool Transformation.withValidation() { return false;}` and override it in `TransactionValidator`

Reply:
I don't want to expose that as part of the contract. The reason is that the validation is somewhat 'hacked' use of transformations. The transformation is forced to have zero outputs. So `Transformation#isValidation` makes no sense, because a proper validator is not a transformation. It is only 'handy' place where to call it, because it has the same context. Otherwise we would have to introduce `validations` and a complete new interface on the level of Repository. We might do it in future, but fow now I'd like to keep it simple.